src/test/java/org/opendatakit/common/persistence/NewQueryResultTest.java

```diff
@@ -0,0 +1,231 @@
+    // We need to sort in order to execute a "cursor-style query"
+    addSort(query);
+
+    int firstPageSize = TEST_ROWS.length / 2;
+    int lastPageSize = TEST_ROWS.length - firstPageSize;
+
+    QueryResult firstPage = query.executeQuery(null, firstPageSize);
+    QueryResumePoint startCursor = firstPage.getResumeCursor();
+    QueryResult lastPage = query.executeQuery(startCursor, lastPageSize);
+
+    assertEquals(lastPageSize, lastPage.getResultList().size());
+    assertFalse(lastPage.hasMoreResults());
+  }
+
+  @Test
+  public void query_executeQuery_backward_cursor_first_page() throws ODKDatastoreException {
+    Query query = ds.createQuery(table, "query_executeQuery_backward_cursor_first_page", user);
+    // We need to sort in order to execute a "cursor-style query"
+    addSort(query);
+
+    QueryResult result = query.executeQuery(null, TEST_ROWS.length / 2);
+
+    assertEquals(true, result.hasMoreResults());
+    assertEquals(TEST_ROWS.length / 2, result.getResultList().size());
+  }
+
+  @Test
+  public void query_executeQuery_backward_cursor_last_page() throws ODKDatastoreException {
+    Query query = ds.createQuery(table, "query_executeQuery_backward_cursor_last_page", user);
+    // We need to sort in order to execute a "cursor-style query"
+    addSort(query);
+
+    int firstPageSize = TEST_ROWS.length / 2;
+    // The backward cursor pivots on the first row, therefore,
+    // pending rows amount is the total of rows minus the one
+    // row we're pivoting on
+    int lastPageSize = TEST_ROWS.length - 1;
+
+    QueryResult firstPage = query.executeQuery(null, firstPageSize);
+    QueryResumePoint startCursor = firstPage.getBackwardCursor();
+    QueryResult lastPage = query.executeQuery(startCursor, lastPageSize);
+
+    assertEquals(lastPageSize, lastPage.getResultList().size());
+    assertFalse(lastPage.hasMoreResults());
+  }
+
+  private static synchronized TestTable createTestTable() throws ODKDatastoreException {
+    System.out.println("Creating the test table");
+    TestTable tablePrototype = new TestTable(ds.getDefaultSchemaName());
+    ds.assertRelation(tablePrototype, user);
+    return tablePrototype;
+  }
+
+  private static void insertRowSets(TestTable table, TestRow row) throws ODKEntityPersistException, ODKOverQuotaException {
+    TestTable element = ds.createEntityUsingRelation(table, user);
+    element.setStringField(TestTable.stringField, row.stringField);
+    element.setLongField(TestTable.integerField, row.integerField.longValue());
+    element.setNumericField(TestTable.doubleField, row.doubleField);
+    element.setDateField(TestTable.dateField, row.dateField);
+    element.setBooleanField(TestTable.booleanField, row.booleanField);
+
+    ds.putEntity(element, user);
+  }
+
+  private static void insertTestFixtures() throws ODKEntityPersistException, ODKOverQuotaException {
+    System.out.println("Inserting " + TEST_ROWS.length + " test rows");

```

Comment:
Is the sysout here intentional?

Reply:
These souts where previously being used and I'm up for removing them.

I think they made sense to have a little feedback while we were inserting a bazillion rows and the test was running for several minutes. Now we insert 200ish rows and it runs on seconds...